sql/scan.go

```diff
@@ -112,31 +106,53 @@ func (n *scanNode) Next() bool {
 			return false
 		}
 
+		var vals valMap
 		if n.primaryKey == nil {
-			// This is the first key for the row, reset our vals map.
-			n.vals = valMap{}
+			// This is the first key for the row, create the vals map in order to
+			// decode the primary key columns.
+			vals = make(valMap, len(n.desc.PrimaryIndex.ColumnIDs))
+			// Reset the qvals map expressions to nil. The expresssion will get
+			// filled in with the column values as we decode the key-value pairs for
+			// the row.
+			for _, e := range n.qvals {
+				e.Expr = nil
+			}
 		}
 
 		var remaining []byte
-		remaining, n.err = decodeIndexKey(n.desc, n.desc.PrimaryIndex, n.vals, kv.Key)
+		remaining, n.err = decodeIndexKey(n.desc, n.desc.PrimaryIndex, vals, kv.Key)
 		if n.err != nil {
 			return false
 		}
-		n.primaryKey = []byte(kv.Key[:len(kv.Key)-len(remaining)])
 
-		// TODO(pmattis): We should avoid looking up the column name by column ID
-		// on every key. One possibility is that we could rewrite col-name
-		// references in expressions to refer to <table-id, col-id> tuples.
+		if n.primaryKey == nil {
+			n.primaryKey = []byte(kv.Key[:len(kv.Key)-len(remaining)])
+
+			// This is the first key for the row, initialize the column values that
+			// are part of the primary key.
+			for id := range n.qvals {
```

Comment:
the keys of `n.qvals` are a superset of the ones of `vals`, right? Just wondering why you didn't just iterate over `vals` (which is likely much smaller).